README.md
# sicinfo-router

error:
Main not found
- necessário

CHANGELOG.md
# sicinfo-router

## [0.1.0] release
### Changes
- altera diretorios de inicializar os app
- altera formato do links do aplicativo para "aplicativo-app"
- altera formato do nome de arquivos json para "aplicativo-app.json"
- altera estrutura do pm2.json: "watch" pode ser caminho absoluto
- altera localização para os diretorio ___dirname___ e ___etc___

``` 
quando os diretorio 'etc' e 'dirname' não são declarados eles recebem o valor de 'cwd'
```

test/12-gate_basic.spec.js
/**
 * https://www.chaijs.com/api/
 * https://dev.to/lucifer1004/test-driven-development-of-an-http-server-with-koajs-25b8
 */
'use strict';

const { join } = require("path");
const chai = require("chai");

const { expect } = chai;

const Gate = require('../src/gate')

describe("Gate Basic - GetEnv", () => {

  const { name, PORT, cwd, etc, cfgs, dirname, test, NODE_ENV } = Gate(
    Object.assign({ name: 'test', PORT: 1, cwd: __dirname, test: true })
  );
  
  it(`Should be name = 'test'`, () => {
    expect(name).to.equal('test');
  })

  it(`Should be cwd = __dirname`, () => {
    expect(cwd).to.equal(__dirname);
  })

  it(`Should be etc = join(__dirname, 'etc')`, () => {
    expect(etc).to.equal(join(__dirname, 'etc'));
  })

  it(`Should be dirname = join(__dirname, 'dist')`, () => {
    expect(dirname).to.equal(join(__dirname, 'dist'));
  })

  it(`Should be cfgs = {}`, () => {
    expect(cfgs).to.be.an('object').to.be.empty;
  })

  it(`Should be PORT = 1`, () => {
    expect(PORT).to.equal(1);
  })

  it(`Should be test = true`, () => {
    expect(test).to.true;
  })

  it(`Should be NODE_ENV = deve`, () => {
    expect(NODE_ENV).to.equal('deve');
  })

});


test/15-auth_basic.spec.js
/**
 * https://www.chaijs.com/api/
 * https://dev.to/lucifer1004/test-driven-development-of-an-http-server-with-koajs-25b8
 * 
 * https://www.chaijs.com/plugins/chai-http/
 */
'use strict';

const chai = require("chai");
const { expect } = chai;

Reflect.set(process.env, 'JWT_SECRET', 'A');

describe("Basic Auth - suite test", () => {

  const Auth = require('../src/auth');


});

test/11-index_basic.spec.js
/**
 * https://www.chaijs.com/api/
 * https://dev.to/lucifer1004/test-driven-development-of-an-http-server-with-koajs-25b8
 * 
 * https://www.chaijs.com/plugins/chai-http/
 */
'use strict';

const chai = require("chai");
require('mocha-sinon');

const { expect } = chai;

describe("SicinfoRouter - Basic suite test (development)", () => {
  const Index = require('../src/index')

  beforeEach(function () {
    this.sinon.stub(console, 'log');
  });

  it('Should be a function', () => {
    expect(typeof (Index)).to.equal('function', 'first');
    expect(typeof (Index().logging)).to.equal('function', 'second');
  })

  it('Should be the message', function () {
    const sicinfoRouter = Index(__filename);
    //@ts-ignore
    expect(console.log.calledOnce).to.be.true;
    //@ts-ignore
    expect(console.log.calledWith(`Loading... ${__filename}`)).to.be.true;
  })

  it(`Should be ContentType = 'content-type'`, () => {
    expect(Index().ContentType).to.equal('content-type')
  })

  it(`Should be TypeTextPlain = 'text/plain; charset=utf-8'`, () => {
    expect(Index().TypeTextPlain).to.equal('text/plain; charset=utf-8')
  })

  it(`Should be TypeApplicationJson = 'application/json; charset=utf-8'`, () => {
    expect(Index().TypeApplicationJson).to.equal('application/json; charset=utf-8')
  })

  it(`Should be TypeApplicationJson = 'application/json; charset=utf-8'`, () => {
    expect(Index().TypeApplicationJson).to.equal('application/json; charset=utf-8')
  })

  it(`Should be ContentTypeTextPlain = ['content-type','text/plain; charset=utf-8']`, () => {
    expect(Index().ContentTypeTextPlain)
      .to.be.an('array').to.include.ordered.members(['content-type', 'text/plain; charset=utf-8']);
  })

  it(`Should be ContentTypeApplicationJson = ['content-type','application/json; charset=utf-8']`, () => {
    expect(Index().ContentTypeApplicationJson)
      .to.be.an('array').to.include.ordered.members(['content-type', 'application/json; charset=utf-8']);
  })

  it(`Should have a property Auth`, () => {
    expect(Index()).to.have.property('Auth')
  })

  it(`Should have a property Router`, () => {
    expect(Index()).to.have.property('Router')
  })

  it(`Should have a property Service`, () => {
    expect(Index()).to.have.property('Service')
  })

  it(`Should have a property etc`, () => {
    expect(Index()).to.have.property('etc')
  })

})

src/auth.js
/**
 * application gate
 * 
 * powered by Moreira in 2019-01-30
 */
'use strict'

const Bearer = 'Bearer';
const Authorization = 'authorization';
const jwt = require('jsonwebtoken');
const { Unauthorized } = require('http-errors');
const expiresIn = 60 * 60; // 1 hora
const JwtSecret = process.env.JWT_SECRET || `${process.env.username}${process.env.HOSTNAME}`

/**
 * @arg { SignArgs } arg
 * @arg { SignOpts } opts
 * @this { Auth }
 */
const _sign = function(arg, opts = {}) {
  const { Bearer, secret } = this;
  // if (opts.algorithm === undefined) opts.algorithm = 'HS256'
  return `${Bearer} ${jwt.sign(arg, secret, opts)}`
}

module.exports =  class {

  /** 
   * @arg { IncomingHttpHeaders } headers
   */
  static verify(headers) {

    const [type, token] = headers.authorization?.split(' ') || [];

    return new Promise((resolve, reject) => {

      if (!type || type !== Bearer || !token) {
        reject(new Unauthorized('jwt invalid'))
      } else jwt.verify(token, this.secret, {}, (err, decoded) => {
        if (err) reject(new Unauthorized(err.message));
        else resolve(decoded);
      });

    })
  }

  /**
   * @arg { IncomingHttpHeaders } [arg0={}]
   * @return { Promise<Token> }
   */
  static validate({ authorization = '' }, requiredAuthorization = false) {

    return new Promise((resolve, reject) => {

      const [type, token] = authorization.split(' ');

      if (!type || type !== Bearer || !token) {
        return requiredAuthorization ? reject(new Unauthorized()) : resolve({})
      }
      
      jwt.verify(token, this.secret, {}, (err, decoded) => {
        if (!err) resolve(decoded || {})
        else if (requiredAuthorization) reject(new Unauthorized(err.message))
        else resolve({})
      });

    });
  }

  /**
   * @arg { SignArgs } arg
   * @arg { SignOpts } opts
   */
  static create(arg = {}, opts = {}) {

    if (undefined === opts.expiresIn) opts.expiresIn = expiresIn;
    if (undefined === arg.stp) arg.stp = 1 * (/** @type { number } */ (opts.expiresIn || 0));
    if (undefined === arg.cnt) arg.cnt = 0;

    return _sign.call(this, arg, opts);
  }

  /** 
   * @arg {SignArgs} arg
   */
  static renew(arg = {}) {

    arg.exp = (Date.now() / 1000 | 0) + (arg.stp || 0)
    arg.cnt = 1 + (arg.cnt || 0)

    return _sign.call(this, arg);
  }

  static get secret() { return JwtSecret }
  static get Authorization() { return Authorization }
  static get Bearer() { return Bearer }

};

src/gate.js
/**
 * module: router
 * 
 * powered by moreira - 2020-11-21
 * 
 */
/* eslint-env: es2020 */
/** 
 * @typedef GateEnv
 * @property {string} [cwd]
 * @property {string} [PWD]
 * @property {string} [PORT]
 * @property {string} [watch]
 * @property {string} [name='']
 * @property {string} [etc='etc']
 * @property {string} [dirname='']
 * @property {boolean} [test=false]
 * @property {Http.Server} [server]
 * @property {Record<string,string>} [cfgs={}]
 * @property {string} [NODE_ENV='development']
 * @property {(...a:[...any]) => void } [logging]
 */ 
'use strict';

const { join } = require('path')
const { readFileSync, statSync } = require('fs');

const Http = require('http');

const { 
  BadRequest, 
  NotFound, 
  InternalServerError, 
  isHttpError 
} = require('http-errors');

/** @return {GateEnv} */
module.exports = (
  /** @type {GateEnv} */ {
    NODE_ENV = 'development',
    name = '',
    cwd = '',
    etc = '',
    dirname = '',
    cfgs = {},
    PORT, PWD,
    watch = "dist",
    server,
    test
  },
  /** @type { function } */ logging
) => {

  NODE_ENV = NODE_ENV.slice(0, 4)

  if (!cwd) cwd = PWD || __dirname
  if (!etc) etc = cwd;
  if (!dirname) dirname = cwd;
  if (!logging) logging = (/** @type {...any} */ ...a) => { NODE_ENV !== 'prod' && console.info(...a) }

  if (!server) server = Http.createServer((req, res) => {
    //@ts-ignore
    logging(`-> ${name}: ${new Date().toISOString()} - HTTP ${req.method} ${req.url}`)

    new Promise((resolve, reject) => {
      
      const [appname, ...url] = (req?.url || '').split('?')[0].replace(/^\/*|\/$/g, '').split('/');
      if (!appname) return reject(new BadRequest('App undefined'));
      if (/\./.test(appname)) return res.writeHead(204).end();
      
      const _dirname = join(dirname, watch.split(',')[0], `app-${appname}`);
      const appjson = join(etc, `app-${appname}.json`);

      if (!cfgs[appjson]) try {

        const stats = statSync(appjson);
        if (!stats.isFile() || stats.size < 12) {
          return reject(new BadRequest('App bad formatted'))
        }

        cfgs[appjson] = Reflect.get(JSON.parse(readFileSync(appjson, { encoding: 'utf8' })), 'main')
        if (!cfgs[appjson]) {
          return reject(new BadRequest('Main undefined'))
        }

      } catch (/** @type {any} */ err) {
        return reject(
          err.code === 'ENOENT' ?
            new NotFound('Main not found') :
            err instanceof SyntaxError ?
              new BadRequest('Main bad formatted') :
              err
        )
      }
      
      resolve([cfgs[appjson], _dirname, url, appname])
    }).then(([main, dirname, url, appname]) => {
      try { 
        new (require(join(dirname, main)))({ 
          req, 
          res, 
          cwd, 
          etc, 
          appname, 
          dirname, 
          url, 
          NODE_ENV 
        });
      }
      catch (/** @type { any } */ err) {
        logging('constructor', err.stack || err);
        throw err.code === 'MODULE_NOT_FOUND' ?
          new NotFound('Service not found') :
          err;
      }
    }).catch(err => {
      if (!isHttpError(err)) {
        err = new InternalServerError(err.message);
      }
      res.statusCode = err.statusCode || err.status;
      res.end(`${err.message || err}`);
    })
  })

  return {
    name, 
    cwd, 
    etc, 
    dirname, 
    cfgs, 
    PORT, 
    test, 
    NODE_ENV,
    server: test ? server : server?.listen(
      PORT, NODE_ENV === 'prod' ? undefined :
      /** @this import('net').Server */ function () {
        const { address, port } = /** @type {*} */ (this.address());
        const lns = [
          `${process.title.split(' ')[0]} ${process.version}`,
          `${new Date().toISOString()}: Listening: (${NODE_ENV})`,
          `-> ${address}:${port}${cwd}`,
          '\n.'.repeat(0)
        ];

        //@ts-ignore
        logging(
          ['-'.repeat(1 * lns.reduce((a, b) => Math.max(a, b.length), 0)), ...lns].join('\n')
        );
      }
    )
  }

}

src/router.js
/** 
 * module: router 
 * 
 * powered by moreira 2020-11-21
 */
'use strict';

const { readFileSync } = require('fs');
const _statuses = require('statuses');
const { join: pathJoin } = require('path');
const qs = require('querystring');
const Auth = require('./auth');

const {
  BadRequest,
  NotFound,
  InternalServerError,
  isHttpError,
  ServiceUnavailable
} = require('http-errors');

const {
  ContentType,
  TypeTextPlain,
  TypeApplicationJson,
  logging
} = require('./index')(__filename);

const { empty: statusesEmpty } = _statuses;

const statuses = (/** @type {string} */ b) => {
  return 1 * /** @type {number} */ (_statuses(b));
}

const Router = module.exports = class {

  /** @arg { Ctx & RouterOptions? } ctx */
  constructor(ctx) {
    /** 
     * @readonly 
     * @private 
     */ 
    this.ctx = Object.assign({ req: {}, res: {} }, ctx);
   
    try {
      const service = new this.service(this);

      this.validate(
        !this.isOptionsMethod && this.headers || {},
        !this.isOptionsMethod && service.requiredAuthorization
      ).then(token => {
        this.ctx.authorization = token;
        return new Promise(resolve => {
          if (!this.isPatchOrPostOrPutMethod) resolve(null);
          else {
            let body = '';
            ctx.req?.setEncoding('utf8')
              .on('data', (/** @type {string} */ chunk) => { body += chunk })
              .on('end', () => { resolve(body ? JSON.parse(body) : {}) })
          }
        });
      }).then((/** @type {Attributes} */ data) => {
        this.ctx.body = data;
        return service.doHttp();
      }).then((arg = { status: 204 }) => {
        if (this.isAuthorized) Reflect.set(
          arg.headers || (arg.headers = {}),
          this.Auth.Authorization,
          this.Auth.renew(this.authorization)
        )
        this.send(arg);        
      }).catch(err => {
        this.send(err);
      })
    
    } catch (/** @type {any} */ err) {
      this.send(
        err.message === 'this.service is not a constructor' ?
          new BadRequest(`Service is not a constructor`) :
          err
      );
    }
  }

  get Auth() {
    return Auth
  }

  get authorization() {
    return this.ctx.authorization || {}
  }

  get isAuthorized() {
    const { sub, iss, iat, exp } = this.authorization;
    return !!(sub && iss && iat && exp);
  }

  /** 
   * @arg { IncomingHttpHeaders } headers }
   * @return { Promise<Token> } 
   */
  validate(headers, required = false) {
    return this.Auth.validate(headers, required)
  }

  /** 
   * @arg { RouterResponse | string } arg 
   */
  send(arg) {

    logging('send', arg && (
      (/** @type {any} */ (arg)).tracer ||
      JSON.stringify(arg)
    ));

    const { res } = this.ctx;
    if (!res.writeHead) return

    if (!arg) arg = new InternalServerError();
    const isError = isHttpError(arg);

    if (typeof (arg) === 'string') arg = isError ? { message: arg } : { data: arg };
    let { status = 0, data, message, headers = {} } = arg;
    const isString = [data, message].some(a => typeof (a) === 'string');

    if (!status) status = statuses(isError ? 'Bad Request' : !data ? 'No Content' : 'Ok')
    if (statusesEmpty[status]) [data, message] = [undefined, undefined]

    if (!!data && !headers[ContentType]) headers[ContentType] = isString ? TypeTextPlain : TypeApplicationJson;
    else if (!data && !!message) data = message;

    res.writeHead(status, headers);
    if (!data) res.end();
    else res.end(isString ? data : JSON.stringify(data));
  }

  get service() {
    return this.ctx.service || (this.ctx.service = this.ctx.Service || (() => {

      const url = this.url, _url = url.splice(0);
      let service;

      while (_url.length && !service) {
        const _name = _url.pop();
        if (!_name) continue;

        const dirServices = pathJoin(this.dirservices, ..._url);

        if (isNaN(parseInt(_name))) try { service = require(pathJoin(dirServices, `${_name}-service`)) }
          catch (err) {
            if (url[0]) try { service = require(pathJoin(dirServices, `${_name}-service`, `${url[0]}-service`)) }
              catch (err) { }
          }

        url.unshift(_name);
        // console.log(135, 'service\n', url, this.url);

      };

      if (service) return service;
      else throw new BadRequest('Service undefined')

    })());
  }

  get NODE_ENV() {
    return this.ctx.NODE_ENV
  }

  /** @return {string} */
  get dirname() {
    return this.ctx.dirname || '.'
  }

  get originalUrl() {
    return this.ctx.req.url || (this.ctx.req.url = '/')
  }

  get url() {
    return this.ctx.url || (this.ctx.url = [])
  }

  get dirservices() {
    return this.ctx.dirservices || (
      this.ctx.dirservices = pathJoin(this.dirname, 'services')
    )
  }

  get headers() {
    return this.ctx.headers || (
      this.ctx.headers = this.ctx.req.headers || {}
    )
  }

  get method() {
    return this.ctx.req.method || '';
  }

  get isGetMethod() {
    return this.method === 'GET'
  }

  get isPutMethod() {
    return this.method === 'PUT'
  }

  get isPostMethod() {
    return this.method === 'POST'
  }

  get isPatchMethod() {
    return this.method === 'PATCH'
  }

  get isDeleteMethod() {
    return this.method === 'DELETE'
  }

  get isOptionsMethod() {
    return this.method === 'OPTIONS'
  }

  get isPatchOrPostOrPutMethod() {
    return this.isPatchMethod || this.isPostMethod || this.isPutMethod
  }

  get query() {
    return this.ctx.query || (this.ctx.query = (a => {
      return a ? qs.parse(a) : undefined;
    })(this.originalUrl.split('?')[1]));
  }

  get body() { 
    return this.ctx.body;
  }

  /** @return { NodeJS.Dict<any> } */
  get DbOptions() {

logging('DbOptions------------------------',
  this.ctx.etc,
  `db-${this.ctx.appname}.json`
)    

    try {
      return JSON.parse(
        readFileSync(
          pathJoin(
            this.ctx.etc,
            `db-${this.ctx.appname}.json`
          )
        ).toString()
      )
    } 
    catch (e) { 
      return {} 
    }
  }

}












src/service.js
/** 
 * module: gate 
 * 
 * powered by moreira 2020-11-21
 */
'use strict';

const { logging } = require('./index')(__filename);

const { 
  MethodNotAllowed, 
  InternalServerError 
} = require('http-errors');

module.exports = class Service {

  /** @arg { Router } router*/
  constructor(router) {
    /** @private */ this.router = router;
  }

  get requiredAuthorization() {
    return false;
  }

  /** @return {Promise<RouterResponse>} */
  doHttp(/** @type {unknown} */ options) {
    return this.router.isGetMethod ? this.doGet(options) :
      this.router.isPostMethod ? this.doPost(options) :
        this.router.isPatchMethod ? this.doPatch(options) :
          this.router.isPutMethod ? this.doPut(options) :
            this.router.isDeleteMethod ? this.doDelete(options) :
              this.router.isOptionsMethod ? this.doOptions(options) :
                Promise.reject(new MethodNotAllowed());
  }

  /** 
   * @param {unknown} [options]
   * @return {Promise<RouterResponse>}
   */
  doGet(options) {
    return this.id ? this.doGetById(this.id, options) :
      this.doc ? this.doGetByDoc(this.doc, options) :
        this.doGetByQuery(options);
  }

  /**
   * @param {string} id
   * @param {unknown} [options]
   * @return {Promise<RouterResponse>} 
   */
  doGetById(id, options) {
    throw new MethodNotAllowed('Method GetById Not Allowed');
  }

  /**
   * @param {Record<string,string>} doc
   * @param {unknown} [options]
   * @return {Promise<RouterResponse>} 
   */
  doGetByDoc(doc, options) {
    throw new MethodNotAllowed('Method GetByDoc Not Allowed');
  }

  /** 
   * @param {unknown} [options]
   * @return {Promise<RouterResponse>}
   */
  doGetByQuery(options) {
    throw new MethodNotAllowed('Method GetByQuery Not Allowed');
  }

  /** 
 * @param {unknown} [options]
 * @return {Promise<RouterResponse>}
 */
  doOptions(options) {
    throw new MethodNotAllowed('Method Options Not Allowed');
  }

  /** 
   * @param {unknown} [options]
   * @return {Promise<RouterResponse>}
   */
  doDelete(options) {
    throw new MethodNotAllowed('Method Delete Not Allowed');
  }

  /** 
 * @param {unknown} [options]
 * @return {Promise<RouterResponse>}
 */
  doPatch(options) {
    throw new MethodNotAllowed('Method Patch Not Allowed');
  }

  /** 
 * @param {unknown} [options]
 * @return {Promise<RouterResponse>}
 */
  doPost(options) {
    throw new MethodNotAllowed('Method Post Not Allowed');
  }

  /** 
 * @param {unknown} [options]
 * @return {Promise<RouterResponse>}
 */
  doPut(options) {
    throw new MethodNotAllowed('Method Put Not Allowed');
  }

  get id() {
    return this.router.url[1];
  }

  get doc() {
    return undefined;
    // return this.router[this.router.isPatchOrPostOrPutMethod ? 'body' : 'query']?.doc$;
  }

  get name() { 
    return this.router.url[0] 
  }

  get body() { 
    return this.router.body 
  }

  get query() { 
    return this.router.query 
  }

  get isAuthorized() { 
    return this.router.isAuthorized 
  }
  
}

typings/index.d.js

/** @typedef { import('jsonwebtoken').SignOptions } SignOpts */
/** @typedef { import('http').IncomingHttpHeaders } IncomingHttpHeaders */
/** @typedef { import('http').Server } HttpServer */
/** @typedef { import('querystring').ParsedUrlQuery } ParsedUrlQuery */

/**
 * @typedef Auth
 * @property { (a:SignArgs) => string } renew
 * @property { (a:SignArgs, b:SignOpts) => string } create
 * @property { (a:IncomingHttpHeaders, b?:boolean) => Promise<Token> } validate
 * @property { string } Authorization
 * @property { string } secret
 * @property { string } Bearer
 */ /**
 * @typedef SignArgs
 * @property { string= } iss
 * @property { string= } sub
 * @property { string= } jti
 * @property { (string | string[])= } aud
 * @property { number= } exp
 * @property { number= } nbf
 * @property { number= } iat
 * @property { number= } stp
 * @property { number= } cnt
 * @property { string= } name
 * @property { string= } api
 */ /**
 * @typedef Token
 * @property { string= } iss
 * @property { string= } sub
 * @property { string= } jti
 * @property { (string | string[])= } aud
 * @property { number= } exp
 * @property { number= } nbf
 * @property { number= } iat
 * @property { string= } name
 * @property { string= } api
 */ /**
 * @typedef AuthError
 * @property { number } code
 * @property { string } message
 * @property { number= } expiredAt
 */

/** @typedef {string|string[]|number|boolean|Date|null|undefined} AttributesTypes */
/** @typedef {Record<string, AttributesTypes>} Attributes */

/** @typedef {import('../src/router')} Router */
/** @typedef {import('../src/service')} Service */

/**
 * @typedef Ctx
 * @property {string} appname
 * @property {string} etc
 * @property {string} NODE_ENV
 * @property {import('http').IncomingMessage} [req]
 * @property {import('http').ServerResponse} [res]
 * @property {string[]} [url]
 * @property {string} [dirname]
 * @property {Service} [Service]
 */ 

/**
 * @typedef RouterResponse
 * @property { number } [status]
 * @property { Attributes|string } [data]
 * @property { Attributes|string } [message]
 * @property { import('http').OutgoingHttpHeaders } [headers]
 */ 

/** 
 * @typedef GateEnv
 * @property {string} [cwd]
 * @property {string} [PWD]
 * @property {string} [PORT]
 * @property {string} [watch='dist']
 * @property {string} [name='']
 * @property {string} [etc='etc']
 * @property {string} [dirname='']
 * @property {boolean} [test=false]
 * @property {HttpServer} [server]
 * @property {Record<string,string>} [cfgs={}]
 * @property {string} [NODE_ENV='development']
 * @property {(...a:[...any]) => void } [logging]
 */ 

  /** 
   * @typedef RouterOptions
   * @property { Attributes } [body]
   * @property { string } dirservices
   * @property { Record<string, any> } req
   * @property { Record<string, any> } res
   * @property { Service } [service]
   * @property { Service } [Service]
   * @property { Token } [authorization]
   * @property { IncomingHttpHeaders } [headers]
   * @property { ParsedUrlQuery } [query]
   */ 


src/index.js
/**
 * module: gate
 * 
 * powered by moreira - 2020-11-21
 * 
 * @author moreira
 * @version 0.1.1
 * 
 * updated 2022.02.17
 * -
 */
'use strict';

exports.ContentType = 'content-type';
exports.TypeTextPlain = 'text/plain; charset=utf-8';
exports.TypeApplicationJson = 'application/json; charset=utf-8';
exports.ContentTypeTextPlain = [exports.ContentType, exports.TypeTextPlain];
exports.ContentTypeApplicationJson = [exports.ContentType, exports.TypeApplicationJson];

Object.defineProperty(exports, 'Auth', { get: () => require('./auth') });
Object.defineProperty(exports, 'Router', { get: () => require('./router') });
Object.defineProperty(exports, 'Service', { get: () => require('./service') });

const logging = (process.env.NODE_ENV?.toLowerCase().slice(0, 4) === 'prod') ?
  (/** @type {string?} */ a) => (/** @type {[...any]} */ ...a) => exports :
  function(/** @type {string?} */ name) {
    return (/** @type {string|number} */ ind, /** @type {[...string]} */ ...args) => {
      if (name) {
        let _log = console.log;
        if (args.length == 1 && typeof (args[0]) === 'function') {
          try { 
            args = [args[0]()] 
          }
          catch (/** @type {any} */ err) { 
            _log = console.warn;
            args = Object.entries(err);
          }
        }
        _log([name].concat(args).map((c, i) => `${i ? '-' : ind} ${c}`).join('\n'));
      }
      return exports;
    }
  }

/** 
 * @param {GateEnv & NodeJS.ProcessEnv} GateEnv
 * @return {GateEnv} 
 */
const gate = ({
  NODE_ENV = 'development',
  name = '',
  cwd = '',
  etc = '',
  dirname = '',
  cfgs = {},
  PORT, PWD,
  watch = '',
  server,
  test = false,
  logging = (/** @type {...any} */ ...args) => { 
    NODE_ENV !== 'prod' && console.info(...args) 
  }
}) => {

  const Http = require('http');
  
  const {
    join: pathJoin
  } = require('path')
  
  const {
    readFileSync,
    statSync
  } = require('fs');
  
  const {
    BadRequest,
    NotFound,
    InternalServerError,
    isHttpError
  } = require('http-errors');

  NODE_ENV = NODE_ENV.slice(0, 4)
  if (!cwd) cwd = PWD || __dirname
  if (!dirname) dirname = cwd;
  if (!watch) watch = pathJoin(dirname, 'dist');
  if (!etc) etc = pathJoin(watch, '..', 'etc');;

  if (!server) server = Http.createServer((req, res) => {

    logging(`-> ${name}: ${new Date().toISOString()} - HTTP ${req.method} ${req.url}`);

    new Promise((resolve, reject) => {
      
      const [appname, ...url] = (req?.url || '').split('?')[0].replace(/^\/*|\/$/g, '').split('/');
      if (!appname) return reject(new BadRequest('App undefined'));
      if (/\./.test(appname)) return res.writeHead(204).end();
      
      const _dirname = pathJoin(watch.split(',')[0], `${appname}-app`);
      const appjson = pathJoin(etc, `${appname}-app.json`);

      if (!cfgs[appjson]) try {

        const stats = statSync(appjson);
        if (!stats.isFile() || stats.size < 12) {
          return reject(new BadRequest('App bad formatted'))
        }

        cfgs[appjson] = Reflect.get(JSON.parse(readFileSync(appjson, { encoding: 'utf8' })), 'main')
        if (!cfgs[appjson]) {
          return reject(new BadRequest('Main undefined'))
        }

      } catch (/** @type {any} */ err) {
        return reject(
          err?.code === 'ENOENT' ?
            new NotFound('Main not found') :
            err instanceof SyntaxError ?
              new BadRequest('Main bad formatted') :
              err
        )
      }
      
      resolve([cfgs[appjson], _dirname, url, appname])
    }).then(([main, dirname, url, appname]) => {
      
      try { 
        
        new (require(pathJoin(dirname, main)))({  
          appname, 
          etc, 
          NODE_ENV ,
          req, 
          res, 
          url, 
          dirname, 
          cwd
        });

      } catch (/** @type { any } */ err) {
        logging('constructor', err.stack || err);
        
        throw err.code === 'MODULE_NOT_FOUND' ?
          new NotFound('Service not found') :
          err;
      }

    }).catch(err => {

      if (!isHttpError(err)) {
        err = new InternalServerError(err.message);
      }
      
      res.statusCode = err.statusCode || err.status;
      res.end(`${err.message || err}`);
    })
  })

  return {
    name,
    cwd,
    etc,
    dirname,
    cfgs,
    PORT,
    test,
    NODE_ENV,
    server: test ? server : server?.listen(
      PORT,
      NODE_ENV === 'prod' ?
        undefined :
        /** @this import('net').Server */
        function () {
          
          const { 
            address, port 
          } = /** @type {*} */ (this.address())

          const args = [
            `${process.title.split(' ')[0]} ${process.version}`,
            `${new Date().toISOString()}: Listening: (${NODE_ENV})`,
            `-> ${address}:${port}${cwd}`,
            '\n.'.repeat(0)
          ]

          const maxLength = 1 * args.reduce((a, b) => Math.max(a, b.length), 0)

          logging(['-'.repeat(maxLength), ...args].join('\n'));
        }
    )
  }
}

exports.gate = gate
exports.etc = gate(process.env)?.etc

/** 
 * @param {string} [name='']
 */
module.exports = function(name = '') {
  return (exports.logging = logging.call(this, name))('Loading...');
}

